src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+                }
+                catch (PipeException e)
+                {
+                    int errorCode = PipeError.GetErrorFromHResult(e.ErrorCode);
+                    if (!IsBrokenPipeError(errorCode))
+                    {
+                        throw;
+                    }
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+                catch (CommunicationException e)
+                {
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+                catch (TimeoutException e)
+                {
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+            }
+            catch (TimeoutException e)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                    new TimeoutException(SR.PipeCloseFailed, e), ExceptionEventType);
+            }
+            catch (PipeException e)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                    ConvertPipeException(SR.PipeCloseFailed, e, TransferOperation.Undefined), ExceptionEventType);
+            }
+            finally
+            {
+                if (shouldCloseHandle)
+                {
+                    CloseHandle(false, null, TransferOperation.Undefined);
+                }
+            }
+        }
+
+        private void CloseHandle(bool abort, string timeoutErrorString, TransferOperation transferOperation)
+        {
+            lock (_readLock)
+            {
+                lock (_writeLock)
+                {
+                    if (_closeState == CloseState.HandleClosed)
+                    {
+                        return;
+                    }
+
+                    _timeoutErrorString = timeoutErrorString;
+                    _timeoutErrorTransferOperation = transferOperation;
+                    _aborted = abort;
+                    _closeState = CloseState.HandleClosed;
+                    _pipe.Close();
+
+                    if (_atEOFEvent != null)
+                    {
+                        _atEOFEvent.Dispose();
+                    }
+                }
+            }
+
+            if (abort)
+            {
+                TraceEventType traceEventType = TraceEventType.Warning;
+
+                // we could be timing out a cached connection
+                if (ExceptionEventType == TraceEventType.Information)
+                {
+                    traceEventType = ExceptionEventType;
+                }
+
+                if (DiagnosticUtility.ShouldTrace(traceEventType))
+                {
+                    TraceUtility.TraceEvent(traceEventType, TraceCode.PipeConnectionAbort, SR.TraceCodePipeConnectionAbort, this);
+                }
+            }
+        }
+
+        private void EnterReadingState()
+        {
+            _inReadingState = true;
```

Comment:
nit: `Debug.Assert(Monitor.IsEntered(_readLock));` is a simple way to validate methods like these are always called in the correct lock